mirgecom/profiling.py

```diff
@@ -0,0 +1,175 @@
+
+        import pytools
+
+        tbl = pytools.Table()
+
+        tbl.add_row(['Function', 'Calls', 'T_min', 'T_avg', 'T_max',
+            'F_min', 'F_avg', 'F_max', 'M_min', 'M_avg', 'M_max', 'BW_avg'])
+
+        from statistics import mean
+
+        for key, value in self.profile_results.items():
+            num_values = len(value)
+
+            times = [v['time'] for v in value]
+            flops = [v['flops'] for v in value]
+            bytes_accessed = [v['bytes_accessed'] for v in value]
+
+            tbl.add_row([key.name, num_values, min(times),
+                int(mean(times)), max(times), min(flops), int(mean(flops)),
+                max(flops), min(bytes_accessed), int(mean(bytes_accessed)),
+                max(bytes_accessed), round(mean(bytes_accessed)/mean(times), 3)])
+
+        if hasattr(pytools.Table, 'github_markdown'):
+            print(tbl.github_markdown())
+        else:
+            print(tbl)
+
+    def call_loopy(self, program, **kwargs) -> dict:
+        program = self.transform_loopy_program(program)
+        assert program.options.return_dict
+        assert program.options.no_numpy
+
+        args_tuple = tuple(
+            (key, value.shape) if hasattr(value, 'shape') else (key, value)
+            for key, value in kwargs.items())
+
+        # Check if we need to get the invoker code to generate integer arguments
+        # N.B.: The invoker code might be different for the same program with
+        # different kwargs
+        if (program not in self.kernel_stats
+                or args_tuple not in self.kernel_stats[program]):
+            executor = program.target.get_kernel_executor(program, self.queue)
+            info = executor.kernel_info(executor.arg_to_dtype_set(kwargs))
+
+            kernel = executor.get_typed_and_scheduled_kernel(
+                executor.arg_to_dtype_set(kwargs))
+
+            data = info.implemented_data_info
+
+            # Generate the wrapper code
+            wrapper = executor.get_wrapper_generator()
+
+            gen = PythonCodeGenerator()
+
+            wrapper.generate_integer_arg_finding_from_shapes(gen, kernel, data)
+            wrapper.generate_integer_arg_finding_from_offsets(gen, kernel, data)
+            wrapper.generate_integer_arg_finding_from_strides(gen, kernel, data)
+
+            types = {}
+            param_dict = {}
+
+            for key, value in kwargs.items():
+                if hasattr(value, 'dtype') and not value.dtype == object:
+                    types[key] = value.dtype
+                param_dict[key] = value
+
+            for key, value in kernel.arg_dict.items():
+                if key not in param_dict:
+                    param_dict[key] = None
+
+            for d in data:
+                if d.name not in param_dict:
+                    param_dict[d.name] = None
+
+            # Run the wrapper code, save argument values in param_dict
+            exec(gen.get(), param_dict)
```

Comment:
If this code were stuck into a function, it
* would execute faster (global var retrievals use `dict` lookups, function-local variables use a more efficient, numbering-based mechanism)
* could be cached.

This is particularly important in a setting where size arguments to kernels can change rapidly.

Reply:
I'm having trouble putting this into a function (related to arg passing). let's discuss in the meeting today.

Reply:
Idea: Put `param_dict` into `args`, make the function return the modified `param_args`.

Reply:
Done.